File: push_swap_prototype/src/internal/solver/sort.go
package solver

import (
	"fmt"
	"os"
	"push_swap_prototype/internal/ops"
	optimizer "push_swap_prototype/internal/peephole"
	"push_swap_prototype/internal/selector"
	"push_swap_prototype/internal/separator"
	"push_swap_prototype/internal/stack"
)


func SolvePushSwap(ps *ops.SortingState) {
	fmt.Printf("[DEBUG] SolvePushSwap: Starting to solve push_swap\n")
	
	if stack.IsSorted(ps.A) { 
		fmt.Printf("[DEBUG] SolvePushSwap: Stack A is already sorted, returning\n")
		return 
	}

	fmt.Printf("[DEBUG] SolvePushSwap: Pushing non-LIS elements into B\n")
	separator.PushNonLISIntoB(ps, true)

	candK := 30
	fmt.Printf("[DEBUG] SolvePushSwap: Starting main loop with candK=%d\n", candK)

	for stack.GetSize(ps.A) > 3 {
		fmt.Printf("[DEBUG] SolvePushSwap: Stack A size = %d, calling PickNearBest\n", stack.GetSize(ps.A))
		pos := selector.PickNearBest(ps, candK)
		fmt.Printf("[DEBUG] SolvePushSwap: Got position from PickNearBest: %+v\n", pos)
		ApplyCombined(ps, pos, true)
	}

	fmt.Printf("[DEBUG] SolvePushSwap: Sorting remaining 3 elements\n")
	SortThree(ps)

	fmt.Printf("[DEBUG] SolvePushSwap: Moving elements back from B to A\n")
	for !stack.IsEmpty(ps.B) {
		fmt.Printf("[DEBUG] SolvePushSwap: Stack B size = %d, calling PickNearBest\n", stack.GetSize(ps.B))
		pos := selector.PickNearBest(ps, candK)
		fmt.Printf("[DEBUG] SolvePushSwap: Got position from PickNearBest: %+v\n", pos)
		ApplyCombined(ps, pos, false) // pa
	}

	fmt.Printf("[DEBUG] SolvePushSwap: Aligning minimum to top\n")
	AlignMinToTop(ps)

	fmt.Printf("[DEBUG] SolvePushSwap: Optimizing operations\n")
	optimizer.OptimizeOps(ps.OpList)
	fmt.Printf("[DEBUG] SolvePushSwap: Finished solving\n")
}

func OpenLogFile() *os.File {
	file, err := os.OpenFile("PickNearBest1.txt", os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		fmt.Printf("Error opening log file: %v\n", err)
		return nil
	}
	return file
}

func LogInput(logFile *os.File, a, b *stack.Stack, candK int, iteration int) {
	if logFile == nil {
		return
	}
	numbersA := stack.ToArray(a)
	numbersB := stack.ToArray(b)
	fmt.Fprintf(logFile, "\nIteration %d input: A=%v B=%v candK=%d\n", iteration, numbersA, numbersB, candK)
}

func LogOutput(logFile *os.File, result selector.Position, iteration int) {
	if logFile == nil {
		return
	}
	fmt.Fprintf(logFile, "Iteration %d output: Position=%+v\n", iteration, result)
}

File: push_swap/src/solver/sort.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   sort.c                                             :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ptison <ptison@student.42prague.com>       +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2025/09/09 18:49:21 by ptison            #+#    #+#             */
/*   Updated: 2025/09/29 22:13:07 by ptison           ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "../../include/solver.h"

bool	solve_push_swap(t_sorting_state *state)
{
	t_position	position;

	if (is_sorted(state->a))
		return (true);
	push_non_lis_into_b(state);
	while (get_size(state->a) > 3)
	{
		position = select_best_move(state);
		if (!position.is_valid)
			return (false);
		apply_combined(state, position, true);
	}
	sort_three(state);
	while (get_size(state->b) > 0)
	{
		position = select_best_move(state);
		if (!position.is_valid)
			return (false);
		apply_combined(state, position, false);
	}
	align_min_to_top(state);
	optimize_ops(&state->operations);
	return (true);
}
